src/hssm/hssm.py

```diff
@@ -598,8 +609,97 @@ def sample(
+            vars_to_keep = set(
+                [var.name for var in getattr(self, "pymc_model").free_RVs]
+            ).intersection(set(list(self._inference_obj["posterior"].data_vars.keys())))
+        else:
+            raise ValueError("Model has not been initialized yet. Something is wrong!")
+
+        setattr(
+            self._inference_obj,
+            "posterior",
+            self._inference_obj["posterior"][list(vars_to_keep)],
+        )
+
         return self.traces
 
+    def vi(
+        self,
+        method: str = "advi",
+        niter: int = 10000,
+        draws=1000,
+        return_idata=True,
+        **vi_kwargs,
+    ) -> pm.Approximation | az.InferenceData:
+        """Perform Variational Inference.
+
+        Parameters
+        ----------
+        niter : int
+            The number of iterations to run the VI algorithm. Defaults to 3000.
+        method : str
+            The method to use for VI. Can be one of "advi" or "fullrank_advi", "svgd",
+            "asvgd".Defaults to "advi".
+        draws : int
+            The number of samples to draw from the posterior distribution.
+            Defaults to 1000.
+        return_idata : bool
+            If True, returns an InferenceData object. Otherwise, returns the
+            approximation object directly. Defaults to True.
+
+        Returns
+        -------
+            pm.Approximation or az.InferenceData: The mean field approximation object.
+        """
+        if self.loglik_kind == "analytical":
+            _logger.warning(
+                "VI is not recommended for the analytical likelihood,"
+                " since gradients can be brittle."
+            )
+        elif self.loglik_kind == "blackbox":
+            raise ValueError(
+                "VI is not supported for blackbox likelihoods, "
+                " since likelihood gradients are needed!"
+            )
+
+        # Run variational inference directly from pymc model
+        with self.pymc_model:
+            self._vi_approx = pm.fit(n=niter, method=method, **vi_kwargs)
+
+        # Get posterior samples from vi-approximation
+        if self._vi_approx is not None:
+            self._inference_obj_vi = self._vi_approx.sample(draws)
+        else:
+            raise ValueError(
+                "VI approximation object has not been initialized yet."
+                " However this check happens after VI is supposed to have run,"
+                " so something is wrong!"
+            )
+
+        # Post-processing
+        if hasattr(self, "pymc_model") and self._inference_obj_vi is not None:
+            vars_to_keep = set(
+                [var.name for var in self.pymc_model.free_RVs]
+            ).intersection(
+                set(list(self._inference_obj_vi["posterior"].data_vars.keys()))
+            )
+        else:
+            raise ValueError("Model has not been initialized yet. Something is wrong!")
```

Comment:
This could be nick-picking, but I never liked the error message "Something is wrong." We should either provide something more informative or leave it out